### PR TITLE
Fix issues with weapons dealers

### LINKED ISSUE #424: Yellow or green star for attached Talon
This is a really minor thing (nitpicking) I noticed in connection to issue #423 

Usually, a weapon with an attachment gets a green star in inventory.
A weapon with attached Talon however gets a yellow star in inventory.

When selling a weapon with attached Talon, the yellow star disappears and is replaced by a green star as can be seen in this screenshot:
yellow-green-star

As I said, it's a very minor thing. Maybe, it can also be fixed with minor effort?

### PR DESCRIPTION
**Pull request by selaux**
Fixes #424 #423.

- Flo now also gets discounts for `IC_WEAPON` class items (according to her merc description)
- Extracted attachment star coloring into own function and check for `IC_LAUNCHER` class instead of talon launcher
- Use extracted function in dealer inventory

### CODE REVIEW

**Comment by JAsmine-ja2** [comment]
Nice fixes!
Imo, other item classes should be added to Flo's discount.

In vanilla JA2, Tony doesn't accept grenades!
However, I usually mod him to accept grenades as well, since it makes sense. He sells 40mm grenades - so why shouldn't he also trade with handgrenades?
Hard to say, if it's an oversight by the devs or a 'real' decision.
But adding grenades (and other classes) to Flo's discount has (almost) no effect on a game as long, as the dealer isn't modded.
So, I'm pro adding more classes.

File: Build/Tactical/ShopKeeper_Interface.cc [hunk 1]
```diff
@@ -2578,8 +2578,7 @@ static UINT32 CalcShopKeeperItemPrice(BOOLEAN fDealerSelling, BOOLEAN fUnitPrice
 		// if it's a GUN or AMMO (but not Launchers, and all attachments and payload is included)
 		switch ( GCM->getItem(usItemID)->getItemClass() )
 		{
-			case IC_GUN:
-			case IC_AMMO:
+			case IC_WEAPON:
```
Comment (lynxlynxlynx) [inline Build/Tactical/ShopKeeper_Interface.cc]: Is the removal of IC_AMMO on purpose? IC_WEAPON doesn't include it. Considering what it includes, the change seems excessive:
`define IC_WEAPON         ( IC_GUN | IC_BLADE | IC_THROWING_KNIFE | IC_LAUNCHER | IC_TENTACLES )`
So I'd prefer to just add IC_LAUNCHER to the cases.

**Comment by chmallon** [comment]
> * Flo now also gets discounts for IC_WEAPON class items (according to her merc description)
This change is wrong.
No item has class IC_WEAPON.
IC_WEAPON is defined as IC_GUN | IC_BLADE | IC_THROWING_KNIFE | IC_LAUNCHER | IC_TENTACLES.
Items only have a single bit set and the test is done with 'case IC_WEAPON:'
So now Flo only gets a discount for ammunition ('case IC_AMMO:') but for no weapon anymore.

> * Extracted attachment star coloring into own function and check for IC_LAUNCHER class instead of talon launcher
> * Use extracted function in dealer inventory
This change misses the case in RenderItemPickupMenu().

**Comment by chmallon** [comment]
> * Flo now also gets discounts for IC_WEAPON class items (according to her merc description)

This change is wrong.
No item hass class IC_WEAPON.
The test is 'case IC_WEAPON:', IC_WEAPON has multiple bits set and items only have a single bit set.
So now Flo gets no discount for weapons anymore, only for ammunition ('case IC_AMMO:').

> * Extracted attachment star coloring into own function and check for IC_LAUNCHER class instead of talon launcher
> * Use extracted function in dealer inventory

This change misses the case in RenderItemPickupMenu().

**Comment by Manuel-K** [comment]
@chmallon You're right. Switch/Case do not work like that.

**Comment by chmallon** [comment]
I think the changes regarding Flo are against the gist of her biography.
It clearly states firearms and also the comments in the code say guns and explicitly not launchers.
Now for example it also includes knives.
Also I think this price discrepancy also appears with other gun attachments like the sniper scope (an IC_MISC).

I should be reverted back to the original state.
It seems this change also was never tested, because it did not work, i.e. Flo got no discount anymore (except for ammunition).

**Pull request merged**